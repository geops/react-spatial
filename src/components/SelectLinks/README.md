#

This demonstrates the use of SelectLinks.

```jsx
import React from  'react'
import SelectLinks from 'react-spatial/components/SelectLinks';

class SelectLinksExample extends React.Component {
  constructor(props) {
    super(props);

    this.options = [
      { label: 'New york', value: 'ny' },
      { label: 'Los angeles', value: 'ls' },
      { label: 'Orlando', value: 'orl' },
    ];

    this.state = { selected: this.options[0] };
  }

  render() {
    return (
      <div className="tm-select-links-example">
        <SelectLinks
          isSelected={opt => this.state.selected.label === opt.label}
          options={this.options}
          onClick={(e, opt) => {
            this.setState({ selected: opt });
          }}
        />
        <div>You selected: {this.state.selected.label}</div>
      </div>
    );
  }
}

<SelectLinksExample />;
```
