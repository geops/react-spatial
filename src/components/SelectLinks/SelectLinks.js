import React from 'react';
import PropTypes from 'prop-types';
import ActionLink from '../ActionLink';

const propTypes = {
  /**
   * CSS class of the container.
   */
  className: PropTypes.string,

  /**
   * CSS class of the current slected link.
   */
  classNameSelected: PropTypes.string,

  /**
   * Values of links.
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      title: PropTypes.string,
    }),
  ).isRequired,

  /**
   * Returns the href attribute.
   */
  getHref: PropTypes.func,

  /**
   * Returns the label to display.
   */
  getLabel: PropTypes.func,

  /**
   * Returns the text added in title attribute.
   */
  getTitle: PropTypes.func,

  /**
   * Returns a boolean indicating whether the object is selected.
   */
  isSelected: PropTypes.func.isRequired,

  /**
   * Function triggered on link's click event.
   *
   * @param {Event} evt The click event object.
   * @param {String} val The value associated to the link clicked.

   */
  onClick: PropTypes.func.isRequired,
};

const defaultProps = {
  className: 'tm-select-links',
  classNameSelected: 'tm-selected',
  getHref: () => '#',
  getLabel: option => option.label,
  getTitle: option => option.title,
};

/**
 * This component displays a list of a element associated to values and
 * behaves like a select element.
 */
const SelectLinks = ({
  className,
  classNameSelected,
  options,
  getHref,
  getLabel,
  getTitle,
  isSelected,
  onClick,
}) => {
  if (!options) {
    return null;
  }
  return (
    <div className={className}>
      {options.map(option => (
        <ActionLink
          key={getLabel(option)}
          href={getHref(option)}
          title={getTitle(option)}
          tabIndex="0"
          className={isSelected(option) ? classNameSelected : undefined}
          onClick={evt => onClick(evt, option)}
        >
          {getLabel(option)}
        </ActionLink>
      ))}
    </div>
  );
};

SelectLinks.propTypes = propTypes;
SelectLinks.defaultProps = defaultProps;

export default SelectLinks;
