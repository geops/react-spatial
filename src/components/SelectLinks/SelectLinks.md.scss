.tm-select-links-example {
  @include a();

  a,
  div {
    padding: 10px;
  }
}
