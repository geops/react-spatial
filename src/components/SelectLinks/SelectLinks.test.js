import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SelectLinks from './SelectLinks';

configure({ adapter: new Adapter() });

describe('SelectLinks', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('displays 3 errors in the console', () => {
      shallow(<SelectLinks />);
      expect(spy).toHaveBeenCalledTimes(3);
    });

    test('matches snapshot', () => {
      const component = renderer.create(<SelectLinks />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when properties are set', () => {
    const onClick = () => {};

    const options = [
      {
        title: 'foo',
        label: 'qux',
      },
      {
        title: 'bar',
        label: 'quux',
      },
      {
        title: 'baz',
        label: 'corge',
      },
    ];

    test('matches snapshot', () => {
      const component = renderer.create(
        <SelectLinks
          isSelected={() => false}
          options={options}
          onClick={onClick}
        />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('matches snapshot when a link is selected', () => {
      const component = renderer.create(
        <SelectLinks
          isSelected={option => option.title === 'bar'}
          options={options}
          onClick={onClick}
        />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('can click a link', () => {
      const fn = jest.fn(onClick);
      const fn2 = jest.fn();
      const evt = {
        preventDefault: fn2,
        stopPropagation: fn2,
      };
      const wrapper = mount(
        <SelectLinks isSelected={() => false} options={options} onClick={fn} />,
      );
      wrapper
        .find('a')
        .first()
        .simulate('click', evt);
      expect(fn).toHaveBeenCalledTimes(1);
    });
  });
});
