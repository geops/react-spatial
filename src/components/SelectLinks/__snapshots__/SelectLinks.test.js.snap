// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectLinks when no properties are set matches snapshot 1`] = `null`;

exports[`SelectLinks when properties are set matches snapshot 1`] = `
<div
  className="tm-select-links"
>
  <a
    aria-label="foo"
    href="#"
    onClick={[Function]}
    title="foo"
  >
    qux
  </a>
  <a
    aria-label="bar"
    href="#"
    onClick={[Function]}
    title="bar"
  >
    quux
  </a>
  <a
    aria-label="baz"
    href="#"
    onClick={[Function]}
    title="baz"
  >
    corge
  </a>
</div>
`;

exports[`SelectLinks when properties are set matches snapshot when a link is selected 1`] = `
<div
  className="tm-select-links"
>
  <a
    aria-label="foo"
    href="#"
    onClick={[Function]}
    title="foo"
  >
    qux
  </a>
  <a
    aria-label="bar"
    className="tm-selected"
    href="#"
    onClick={[Function]}
    title="bar"
  >
    quux
  </a>
  <a
    aria-label="baz"
    href="#"
    onClick={[Function]}
    title="baz"
  >
    corge
  </a>
</div>
`;
